mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala

```diff
@@ -378,6 +378,27 @@ object PowerIterationClustering extends Logging {
       logInfo(s"$msgPrefix: delta = $delta.")
       diffDelta = math.abs(delta - prevDelta)
       logInfo(s"$msgPrefix: diff(delta) = $diffDelta.")
+
+      if (math.abs(diffDelta) < tol) {
+        /**
+         * Power Iteration fails to converge if absolute value of top 2 eigen values are equal,
+         * but with opposite sign. The resultant vector flip-flops between two vectors.
+         * We should give an exception, if we detect the failure of the convergence of the
+         * power iteration
+         */
+
+        // Rayleigh quotient = x^tAx / x^tx
+        val xTAx = curG.joinVertices(v) {
+          case (_, x, y) => x * y
+        }.vertices.values.sum()
+        val xTx = curG.vertices.mapValues(x => x * x).values.sum()
+        val Rayleigh = xTAx / xTx
+
+        if (math.abs(norm - math.abs(Rayleigh)) > tol) {
+          throw new SparkException(s"Power Iteration fail to converge, delta = ${delta}," +
```

Comment:
Good question. Is it result meaningful at all if it failed to converge? or is definitely an error? you would know better. Right now it doesn't fail, so this would be a behavior change. It could be correct if the result is just never usable in this case.

Reply:
The result may be partially correct. Refer UT: even though both the clusters are well separable, any standard clustering algorithm would correctly assign the clusters, but due to the drawback of PIC, it gives incorrect result. 

So, if we log an error, it is up to user to decide, whether to take the results from the PIC or use some other algorithm.